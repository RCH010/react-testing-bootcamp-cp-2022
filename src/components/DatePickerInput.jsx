import React from 'react';

import TextField from '@mui/material/TextField';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFns';
import { LocalizationProvider } from '@mui/x-date-pickers/LocalizationProvider';
import { DatePicker } from '@mui/x-date-pickers/DatePicker';

export const DatePickerInput = ({
  value = new Date(),
  setValue = () => {},
  validDate = true,
  setValidDate = () => {},
  maxDate = new Date(),
  minDate = new Date(0),
}) => {
  const handleDateChange = (selectedDate) => {
    let dateNumber = selectedDate && Date.parse(selectedDate);
    const max = Date.parse(maxDate);
    const min = Date.parse(minDate);

    setValidDate(!isNaN(dateNumber) && dateNumber < max && dateNumber > min);
    setValue(selectedDate);
  };

  return (
    <LocalizationProvider dateAdapter={AdapterDateFns}>
      <DatePicker
        maxDate={maxDate}
        minDate={minDate}
        label='Pick a date'
        value={value}
        onChange={handleDateChange}
        renderInput={(params) => (
          <TextField
            {...params}
            size='small'
            helperText={!validDate ? 'Please provide a valid date' : ''}
          />
        )}
      />
    </LocalizationProvider>
  );
};
