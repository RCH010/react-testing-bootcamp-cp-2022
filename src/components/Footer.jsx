import { Box, Typography } from '@mui/material';
import React from 'react';

export const Footer = () => {
  return (
    <Box component='footer' sx={{ width: '100%', p: 4 }}>
      <Typography variant='body2' textAlign='center'>
        Project created during Wizeline Academy React Testing Bootcamp
      </Typography>
    </Box>
  );
};
