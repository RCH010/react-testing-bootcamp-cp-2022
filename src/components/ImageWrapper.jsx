import React from 'react';
import PropTypes from 'prop-types';
import { Box, Grid, Skeleton, Typography } from '@mui/material';

const LinesOfSkeletons = () =>
  [...Array(6).keys()].map((index) => <Skeleton variant='text' key={index} />);

export const ImageWrapper = ({
  isLoading,
  title,
  currentDate,
  imageSRC,
  description,
}) => {
  return (
    <Grid
      container
      alignItems='center'
      spacing={4}
      justifyContent='space-between'
      direction={{ xs: 'column', md: 'row' }}>
      <Grid item xs={12} md={6} sx={{ width: '100%' }}>
        <Typography variant='h4' component='h2'>
          {isLoading ? <Skeleton /> : title}
        </Typography>
        <Typography variant='body2'>
          {isLoading ? <Skeleton /> : currentDate}
        </Typography>
        <Box sx={{ '& img': { width: '100%' } }}>
          {isLoading ? (
            <Skeleton variant='rectangular' height='25em' />
          ) : (
            <img src={imageSRC} alt={title} />
          )}
        </Box>
      </Grid>
      <Grid item xs={12} md={6} sx={{ width: '100%' }}>
        <Typography variant='body1'>
          {isLoading ? <LinesOfSkeletons /> : description}
        </Typography>
      </Grid>
    </Grid>
  );
};

ImageWrapper.propTypes = {
  title: PropTypes.string,
  currentDate: PropTypes.string,
  imageSRC: PropTypes.string,
  description: PropTypes.string,
};
