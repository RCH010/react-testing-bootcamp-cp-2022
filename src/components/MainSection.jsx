import React, { useEffect, useState } from 'react';

import { DatePickerInput } from './DatePickerInput';
import { ImageWrapper } from './ImageWrapper';
import { fetchPictureOfTheDay } from '../api/apod';

import { Button, Stack, Typography, Box } from '@mui/material';

export const MainSection = () => {
  const [isLoading, setIsLoading] = useState(true);
  const [apodData, setApodData] = useState(null);
  const [errorMessage, setErrorMessage] = useState(null);
  const [apodDate, setapodDate] = useState(null);
  const [validDate, setValidDate] = useState(new Date());

  useEffect(() => {
    const fetchImage = async () => {
      if (isLoading) {
        try {
          const response = await fetchPictureOfTheDay(apodDate);
          setApodData(response);
          setIsLoading(false);
          setErrorMessage(null);
        } catch (error) {
          setErrorMessage(error.msg || 'There was an error, please try again');
          setIsLoading(false);
        }
      }
    };
    fetchImage();
  }, [isLoading]);

  const onSearchNewDate = (e) => {
    e.preventDefault();
    if (validDate) {
      setIsLoading(true);
    }
  };

  return (
    <Box sx={{ px: 4, flexGrow: 1 }}>
      <Stack
        component='form'
        onSubmit={onSearchNewDate}
        direction='row'
        spacing={4}
        justifyContent='flex-end'
        sx={{ mb: 4 }}>
        <DatePickerInput
          value={apodDate}
          setValue={setapodDate}
          validDate={validDate}
          setValidDate={setValidDate}
        />
        <Button
          variant='contained'
          type='submit'
          disabled={!validDate}
          size='small'>
          Search
        </Button>
      </Stack>
      {errorMessage ? (
        <Box>
          <Typography variant='h3' component='h2'>
            Oh no!
          </Typography>
          <Typography>{errorMessage}</Typography>
        </Box>
      ) : (
        <ImageWrapper
          isLoading={isLoading}
          title={apodData?.title}
          description={apodData?.explanation}
          imageSRC={apodData?.url}
          currentDate={apodData?.date}
        />
      )}
    </Box>
  );
};
