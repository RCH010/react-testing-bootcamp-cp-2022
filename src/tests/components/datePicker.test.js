import { render, screen } from '@testing-library/react';

import { DatePickerInput } from '../../components/DatePickerInput';

describe('Tests on datePicker component', () => {
  test('should match snapshot', () => {
    const { container } = render(
      <DatePickerInput value={new Date('2/1/20')} />
    );
    expect(container).toMatchSnapshot();
  });

  test('should have todays date as default value', async () => {
    const defaultDate = new Date();
    render(<DatePickerInput value={defaultDate} />);
    const dateInput = screen.getByLabelText(/Pick a date/i);
    const inputDate = new Date(dateInput.value);
    expect(inputDate.toLocaleDateString()).toBe(
      defaultDate.toLocaleDateString()
    );
  });

  test('should display error message', () => {
    render(<DatePickerInput validDate={false} />);

    const errorMsg = screen.getByText(/Please provide a valid date/i);
    expect(errorMsg).toBeInTheDocument();
  });
});
