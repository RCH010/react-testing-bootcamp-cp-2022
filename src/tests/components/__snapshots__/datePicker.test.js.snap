// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests on datePicker component should match snapshot 1`] = `
<div>
  <div
    class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeSmall MuiInputLabel-outlined MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      for=":r0:"
      id=":r0:-label"
    >
      Pick a date
    </label>
    <div
      class="MuiOutlinedInput-root MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
    >
      <input
        aria-invalid="false"
        aria-label="Choose date, selected date is Feb 1, 2020"
        aria-readonly="true"
        class="MuiOutlinedInput-input MuiInputBase-input MuiInputBase-inputSizeSmall css-1n4twyu-MuiInputBase-input-MuiOutlinedInput-input"
        id=":r0:"
        readonly=""
        type="text"
        value="02/01/2020"
      />
      <fieldset
        aria-hidden="true"
        class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
      >
        <legend
          class="css-1in441m"
        >
          <span>
            Pick a date
          </span>
        </legend>
      </fieldset>
    </div>
  </div>
</div>
`;
